### ISSUE TITLE
Improvements of std_examples/timer/src/main.rs

### ISSUE THREAD

**Issue by prisme60** [apollolabsdev/ESP32C3#2]
Hello Omar,

I read your blog on the [ESP32 Standard Library Embedded Rust: Timers](https://apollolabsblog.hashnode.dev/esp32-standard-library-embedded-rust-timers).

I was trying to factorize the code like following : 
```rust
use esp_idf_hal::gpio::*;
use esp_idf_hal::peripherals::Peripherals;
use esp_idf_hal::timer::config::Config;
use esp_idf_hal::timer::TimerDriver;

fn main() {
    esp_idf_sys::link_patches();

    let peripherals = Peripherals::take().unwrap();

    // Configure and Initialize Timer Drivers
    let config = Config::new();
    let mut timer1 = TimerDriver::new(peripherals.timer00, &config).unwrap();
    let mut timer2 = TimerDriver::new(peripherals.timer10, &config).unwrap();

    // Configure Pins that Will Read the Square Wave as Inputs
    let pin1 = PinDriver::input(peripherals.pins.gpio0.downgrade_input()).unwrap();
    let pin2 = PinDriver::input(peripherals.pins.gpio1.downgrade_input()).unwrap();

    // Declare and Init Variables that will Track Pin level
    let mut pin1_old_level: Level = Level::Low;
    let mut pin2_old_level: Level = Level::High;

    // Set Counter Start Value to Zero
    timer1.set_counter(0_u64).unwrap();
    timer2.set_counter(0_u64).unwrap();

    // Enable Counter
    timer1.enable(true).unwrap();
    timer2.enable(true).unwrap();

    loop {
        pin1_old_level = measure_pin(&mut timer1, &pin1, 1, pin1_old_level);
        pin2_old_level = measure_pin(&mut timer2, &pin2, 2, pin2_old_level);
    }
}

fn measure_pin<T: Pin, MODE: esp_idf_hal::gpio::InputMode>(timer: &mut TimerDriver, pin: &PinDriver<T, MODE>, index: u8, pin_previous_level: Level) -> Level {
    // Get Level of pin
    let pin_current_level = pin.get_level();

    if pin_current_level != pin_previous_level {
        match pin_current_level {
            // If pin level changed from Low to High then reset count
            Level::High => timer.set_counter(0).unwrap(),
            // If pin level changed from High to Low then capture count
            Level::Low => {
                let count: u64 = timer.counter().unwrap();
                // Calculate and Print Out the Pulse Width
                // Clock Frequency is 1 MHz According to Code
                println!("Sq Wave {} Pulse Width is {:.1}ms", index, count / 1000);
            }
        }
    }

    pin_current_level
}
```
I test it with [Wokwi](https://wokwi.com/projects/372767628999403521) and the time accuracy is more precise (and take less emulated CPU).

First, I wonder why, then I understand that you print several time the same count at each iteration loop.
![image](https://github.com/apollolabsdev/ESP32C3/assets/1153403/e7e745c7-c5d3-410c-aed8-031d6cdbd4e1)

In my revised code, I only display the count, when the new measure is available. So the display of the message is triggered by the pin state change. Because Wave 1 pulse is 25ms and Wave 2 pulse is 10ms. We have more wave 2 pulse messages than wave 1 pulse messages.
![image](https://github.com/apollolabsdev/ESP32C3/assets/1153403/700e7750-7127-43a7-a064-05c3e15eb38b)

**Comment by prisme60** [comment]
I have update [Wokwi](https://wokwi.com/projects/372767628999403521) with another improvement:
I saw that you put the following code for displaying pulse width :
```rust
println!("Sq Wave 1 Pulse Width is {:.1}ms", count1 / 1000);
```
Because of the `{:.1}`, you expect to have a decimal digit, but `count1/1000` gives an integer result.

There are several possibilities in order to get a decimal digit:
- use float (if hardware doesn't natively support float, it will use soft float, that can be slow and increase flash footprint)
```rust
println!("Sq Wave 1 Pulse Width is {:.1}ms", count1 as f32 / 1000f32);
```
- compute entire part and decimal part:
```rust
let integer_part  = count1 / 1000;
let decimal_part  = (count1 % 1000) / 100;
println!("Sq Wave 1 Pulse Width is {}.{}ms", integer_part, decimal_part);
```

- use [rust_decimal crate](https://crates.io/crates/rust_decimal)
```rust
let decimal_count = rust_decimal::Decimal::new(count1 as i64, 3); // decimal of 3 digits
println!("Sq Wave 1 Pulse Width is {:.1}ms", decimal_count);
```

![image](https://github.com/apollolabsdev/ESP32C3/assets/1153403/f1c3a49c-c2be-49b3-a00b-59d95ce4abd8)

Full code :
```rust
use esp_idf_hal::gpio::*;
use esp_idf_hal::peripherals::Peripherals;
use esp_idf_hal::timer::config::Config;
use esp_idf_hal::timer::TimerDriver;

fn main() {
    esp_idf_sys::link_patches();

    let peripherals = Peripherals::take().unwrap();

    // Configure and Initialize Timer Drivers
    let config = Config::new();
    let mut timer1 = TimerDriver::new(peripherals.timer00, &config).unwrap();
    let mut timer2 = TimerDriver::new(peripherals.timer10, &config).unwrap();

    // Configure Pins that Will Read the Square Wave as Inputs
    let pin1 = PinDriver::input(peripherals.pins.gpio0.downgrade_input()).unwrap();
    let pin2 = PinDriver::input(peripherals.pins.gpio1.downgrade_input()).unwrap();

    // Declare and Init Variables that will Track Pin level
    let mut pin1_old_level: Level = Level::Low;
    let mut pin2_old_level: Level = Level::High;

    // Set Counter Start Value to Zero
    timer1.set_counter(0_u64).unwrap();
    timer2.set_counter(0_u64).unwrap();

    // Enable Counter
    timer1.enable(true).unwrap();
    timer2.enable(true).unwrap();

    loop {
        pin1_old_level = measure_pin(&mut timer1, &pin1, 1, pin1_old_level);
        pin2_old_level = measure_pin(&mut timer2, &pin2, 2, pin2_old_level);
    }
}

fn measure_pin<T: Pin, MODE: esp_idf_hal::gpio::InputMode>(timer: &mut TimerDriver, pin: &PinDriver<T, MODE>, index: u8, pin_previous_level: Level) -> Level {
    // Get Level of pin
    let pin_current_level = pin.get_level();

    if pin_current_level != pin_previous_level {
        match pin_current_level {
            // If pin level changed from Low to High then reset count
            Level::High => timer.set_counter(0).unwrap(),
            // If pin level changed from High to Low then capture count
            Level::Low => {
                let count: u64 = timer.counter().unwrap();

                //**********integer version (no decimal digit)************
                //println!("Sq Wave {} Pulse Width is {}ms", index, count);
                //**********float version************

                //**********float version************
                //println!("Sq Wave {} Pulse Width is {:.1}ms", index, count as f32 / 1000f32);
                //**********float version************

                //**********remainder version********
                let integer_part  = count / 1000;
                let decimal_part  = (count % 1000) / 100;
                // Calculate and Print Out the Pulse Width
                // Clock Frequency is 1 MHz According to Code
                println!("Sq Wave {} Pulse Width is {}.{}ms", index, integer_part, decimal_part);
                //**********remainder version********

                //*****rust_decimal crate version********
                //let decimal_count = rust_decimal::Decimal::new(count as i64, 3); // decimal of 3 digits
                //println!("Sq Wave {} Pulse Width is {:.1}ms", index, decimal_count);
                //*****rust_decimal crate version********
            }
        }
    }

    pin_current_level
}
```

**Comment by apollolabsdev** [comment]
Thank you for the suggestions! The blog post was updated to include the suggested enhancements for the code. Regarding using a function, while the implementation is more efficient, I try to avoid it in embedded at least since getting the parameter types correctly might be cumbersome. As for the floating point, it was an artifact from an earlier implementation I was experimenting with. I figured I didn't need that much accuracy and stuck to integer math as it's lighter weight.
